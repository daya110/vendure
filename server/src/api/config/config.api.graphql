type Query {
  config: Config!
}

type Config {
    customFields: JSON
}
