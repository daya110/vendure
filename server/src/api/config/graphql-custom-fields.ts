import { buildSchema, extendSchema, parse, printSchema } from 'graphql';

import { CustomFieldConfig, CustomFields, CustomFieldType } from '../../../../shared/shared-types';
import { assertNever } from '../../../../shared/shared-utils';

/**
 * Given a CustomFields config object, generates an SDL string extending the built-in
 * types with a customFields property for all entities, translations and inputs for which
 * custom fields are defined.
 */
export function addGraphQLCustomFields(typeDefs: string, customFieldConfig: CustomFields): string {
    const schema = buildSchema(typeDefs);

    let customFieldTypeDefs = '';

    if (!schema.getType('JSON')) {
        customFieldTypeDefs += `
            scalar JSON
        `;
    }

    for (const entityName of Object.keys(customFieldConfig)) {
        const customEntityFields = customFieldConfig[entityName as keyof CustomFields] || [];

        const localeStringFields = customEntityFields.filter(field => field.type === 'localeString');
        const nonLocaleStringFields = customEntityFields.filter(field => field.type !== 'localeString');

        if (schema.getType(entityName)) {
            if (customEntityFields.length) {
                customFieldTypeDefs += `
                    type ${entityName}CustomFields {
                        ${mapToFields(customEntityFields, getGraphQlType)}
                    }

                    extend type ${entityName} {
                        customFields: ${entityName}CustomFields
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend type ${entityName} {
                        customFields: JSON
                    }
                `;
            }
        }

        if (localeStringFields.length && schema.getType(`${entityName}Translation`)) {
            customFieldTypeDefs += `
                    type ${entityName}TranslationCustomFields {
                         ${mapToFields(localeStringFields, getGraphQlType)}
                    }

                    extend type ${entityName}Translation {
                        customFields: ${entityName}TranslationCustomFields
                    }
                `;
        }

        if (schema.getType(`Create${entityName}Input`)) {
            if (nonLocaleStringFields.length) {
                customFieldTypeDefs += `
                    input Create${entityName}CustomFieldsInput {
                       ${mapToFields(nonLocaleStringFields, getGraphQlType)}
                    }

                    extend input Create${entityName}Input {
                        customFields: Create${entityName}CustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                   extend input Create${entityName}Input {
                       customFields: JSON
                   }
               `;
            }
        }

        if (schema.getType(`Update${entityName}Input`)) {
            if (nonLocaleStringFields.length) {
                customFieldTypeDefs += `
                    input Update${entityName}CustomFieldsInput {
                       ${mapToFields(nonLocaleStringFields, getGraphQlType)}
                    }

                    extend input Update${entityName}Input {
                        customFields: Update${entityName}CustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend input Update${entityName}Input {
                        customFields: JSON
                    }
                `;
            }
        }

        if (customEntityFields.length && schema.getType(`${entityName}SortParameter`)) {
            customFieldTypeDefs += `
                    extend input ${entityName}SortParameter {
                         ${mapToFields(customEntityFields, () => 'SortOrder')}
                    }
                `;
        }

        if (customEntityFields.length && schema.getType(`${entityName}FilterParameter`)) {
            customFieldTypeDefs += `
                    extend input ${entityName}FilterParameter {
                         ${mapToFields(customEntityFields, getFilterOperator)}
                    }
                `;
        }

        if (localeStringFields && schema.getType(`${entityName}TranslationInput`)) {
            if (localeStringFields.length) {
                customFieldTypeDefs += `
                    input ${entityName}TranslationCustomFieldsInput {
                        ${mapToFields(localeStringFields, getGraphQlType)}
                    }

                    extend input ${entityName}TranslationInput {
                        customFields: ${entityName}TranslationCustomFieldsInput
                    }
                `;
            } else {
                customFieldTypeDefs += `
                    extend input ${entityName}TranslationInput {
                        customFields: JSON
                    }
                `;
            }
        }
    }

    return printSchema(extendSchema(schema, parse(customFieldTypeDefs)));
}

type GraphQLFieldType = 'String' | 'Int' | 'Float' | 'Boolean' | 'ID';

/**
 * Maps an array of CustomFieldConfig objects into a string of SDL fields.
 */
function mapToFields(fieldDefs: CustomFieldConfig[], typeFn: (fieldType: CustomFieldType) => string): string {
    return fieldDefs.map(field => `${field.name}: ${typeFn(field.type)}`).join('\n');
}

function getFilterOperator(type: CustomFieldType): string {
    switch (type) {
        case 'datetime':
            return 'DateOperators';
        case 'string':
        case 'localeString':
            return 'StringOperators';
        case 'boolean':
            return 'BooleanOperators';
        case 'int':
        case 'float':
            return 'NumberOperators';
        default:
            assertNever(type);
    }
    return 'String';
}

function getGraphQlType(type: CustomFieldType): GraphQLFieldType {
    switch (type) {
        case 'string':
        case 'datetime':
        case 'localeString':
            return 'String';
        case 'boolean':
            return 'Boolean';
        case 'int':
            return 'Int';
        case 'float':
            return 'Float';
        default:
            assertNever(type);
    }
    return 'String';
}
