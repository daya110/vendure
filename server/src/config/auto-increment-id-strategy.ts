import { IntegerIdStrategy } from './entity-id-strategy';

/**
 * An id strategy which simply uses auto-increment integers as primary keys
 * for all entities.
 */
export class AutoIncrementIdStrategy implements IntegerIdStrategy {
    primaryKeyType: 'increment';
    decodeId(id: string): number {
        return +id;
    }
    encodeId(primaryKey: number): string {
        return primaryKey.toString();
    }
}
