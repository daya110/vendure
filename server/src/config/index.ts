export * from './asset-naming-strategy/asset-naming-strategy';
export * from './asset-naming-strategy/default-asset-naming-strategy';
export * from './asset-preview-strategy/asset-preview-strategy';
export * from './asset-storage-strategy/asset-storage-strategy';
export * from './email/email-options';
export * from './email/email-transport-options';
export * from './entity-id-strategy/auto-increment-id-strategy';
export * from './entity-id-strategy/entity-id-strategy';
export * from './entity-id-strategy/uuid-id-strategy';
export * from './order-merge-strategy/order-merge-strategy';
export * from './payment-method/fakepal-payment-method-config';
export * from './payment-method/gripe-payment-method-config';
export * from './payment-method/payment-method-handler';
export * from './promotion/default-promotion-actions';
export * from './promotion/default-promotion-conditions';
export * from './promotion/promotion-action';
export * from './promotion/promotion-condition';
export * from './shipping-method/default-shipping-calculator';
export * from './shipping-method/default-shipping-eligibility-checker';
export * from './shipping-method/shipping-calculator';
export * from './shipping-method/shipping-eligibility-checker';
export * from './vendure-config';
export * from './vendure-plugin/vendure-plugin';
