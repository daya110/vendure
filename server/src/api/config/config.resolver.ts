import { Query, Resolver } from '@nestjs/graphql';

import { getConfig, VendureConfig } from '../../config/vendure-config';
import { ConfigService } from '../../service/config.service';

@Resolver('Config')
export class ConfigResolver {
    constructor(private configService: ConfigService) {}

    /**
     * Exposes a subset of the VendureConfig which may be of use to clients.
     */
    @Query()
    config(): Partial<VendureConfig> {
        return {
            customFields: this.configService.customFields,
        };
    }
}
