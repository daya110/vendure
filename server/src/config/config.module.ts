import { Module } from '@nestjs/common';

import { I18nService } from '../i18n/i18n.service';

import { ConfigService } from './config.service';

@Module({
    providers: [ConfigService],
    exports: [ConfigService],
})
export class ConfigModule {}
