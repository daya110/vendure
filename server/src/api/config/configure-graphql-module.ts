import { DynamicModule } from '@nestjs/common';
import { GqlModuleOptions, GraphQLModule, GraphQLTypesLoader } from '@nestjs/graphql';
import { GraphQLUpload } from 'apollo-server-core';
import { extendSchema, printSchema } from 'graphql';
import { GraphQLDateTime } from 'graphql-iso-date';
import GraphQLJSON from 'graphql-type-json';

import { ConfigModule } from '../../config/config.module';
import { ConfigService } from '../../config/config.service';
import { I18nModule } from '../../i18n/i18n.module';
import { I18nService } from '../../i18n/i18n.service';
import { getPluginAPIExtensions } from '../../plugin/plugin-utils';
import { ApiSharedModule } from '../api-internal-modules';
import { IdCodecService } from '../common/id-codec.service';
import { IdEncoderExtension } from '../middleware/id-encoder-extension';
import { TranslateErrorExtension } from '../middleware/translate-errors-extension';

import { generateListOptions } from './generate-list-options';
import { addGraphQLCustomFields } from './graphql-custom-fields';

export interface GraphQLApiOptions {
    apiType: 'shop' | 'admin';
    typePaths: string[];
    apiPath: string;
    // tslint:disable-next-line:ban-types
    resolverModule: Function;
}

/**
 * Dynamically generates a GraphQLModule according to the given config options.
 */
export function configureGraphQLModule(
    getOptions: (configService: ConfigService) => GraphQLApiOptions,
): DynamicModule {
    return GraphQLModule.forRootAsync({
        useFactory: (
            configService: ConfigService,
            i18nService: I18nService,
            idCodecService: IdCodecService,
            typesLoader: GraphQLTypesLoader,
        ) => {
            return createGraphQLOptions(
                i18nService,
                configService,
                idCodecService,
                typesLoader,
                getOptions(configService),
            );
        },
        inject: [ConfigService, I18nService, IdCodecService, GraphQLTypesLoader],
        imports: [ConfigModule, I18nModule, ApiSharedModule],
    });
}

function createGraphQLOptions(
    i18nService: I18nService,
    configService: ConfigService,
    idCodecService: IdCodecService,
    typesLoader: GraphQLTypesLoader,
    options: GraphQLApiOptions,
): GqlModuleOptions {
    // Prevent `Type "Node" is missing a "resolveType" resolver.` warnings.
    // See https://github.com/apollographql/apollo-server/issues/1075
    const dummyResolveType = {
        __resolveType() {
            return null;
        },
    };

    return {
        path: '/' + options.apiPath,
        typeDefs: createTypeDefs(options.apiType),
        include: [options.resolverModule],
        resolvers: {
            JSON: GraphQLJSON,
            DateTime: GraphQLDateTime,
            Node: dummyResolveType,
            PaginatedList: dummyResolveType,
            Upload: GraphQLUpload || dummyResolveType,
            SearchResultPrice: {
                __resolveType(value: any) {
                    return value.hasOwnProperty('value') ? 'SinglePrice' : 'PriceRange';
                },
            },
        },
        uploads: {
            maxFileSize: configService.assetOptions.uploadMaxFileSize,
        },
        playground: true,
        debug: true,
        context: (req: any) => req,
        extensions: [
            () => new TranslateErrorExtension(i18nService),
            () => new IdEncoderExtension(idCodecService),
        ],
        // This is handled by the Express cors plugin
        cors: false,
    };

    /**
     * Generates the server's GraphQL schema by combining:
     * 1. the default schema as defined in the source .graphql files specified by `typePaths`
     * 2. any custom fields defined in the config
     * 3. any schema extensions defined by plugins
     */
    function createTypeDefs(apiType: 'shop' | 'admin'): string {
        const customFields = configService.customFields;
        const typeDefs = typesLoader.mergeTypesByPaths(...options.typePaths);
        let schema = generateListOptions(typeDefs);
        schema = addGraphQLCustomFields(schema, customFields);
        const pluginSchemaExtensions = getPluginAPIExtensions(configService.plugins, apiType).map(
            e => e.schema,
        );

        for (const documentNode of pluginSchemaExtensions) {
            schema = extendSchema(schema, documentNode);
        }
        return printSchema(schema);
    }
}
