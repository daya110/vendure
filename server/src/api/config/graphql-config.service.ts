import { DynamicModule } from '@nestjs/common';
import { GqlModuleOptions, GraphQLModule, GraphQLTypesLoader } from '@nestjs/graphql';
import { GraphQLUpload } from 'apollo-server-core';
import { extendSchema, printSchema } from 'graphql';
import { GraphQLDateTime } from 'graphql-iso-date';
import GraphQLJSON from 'graphql-type-json';

import { notNullOrUndefined } from '../../../../shared/shared-utils';
import { ConfigModule } from '../../config/config.module';
import { ConfigService } from '../../config/config.service';
import { I18nModule } from '../../i18n/i18n.module';
import { I18nService } from '../../i18n/i18n.service';
import { TranslateErrorExtension } from '../middleware/translate-errors-extension';

import { generateListOptions } from './generate-list-options';
import { addGraphQLCustomFields } from './graphql-custom-fields';

export interface GraphQLApiOptions {
    typePaths: string[];
    apiPath: string;
    // tslint:disable-next-line:ban-types
    resolverModule: Function;
}

/**
 * Dynamically generates a GraphQLModule according to the given config options.
 */
export function configureGraphQLModule(
    getOptions: (configService: ConfigService) => GraphQLApiOptions,
): DynamicModule {
    return GraphQLModule.forRootAsync({
        useFactory: (
            configService: ConfigService,
            i18nService: I18nService,
            typesLoader: GraphQLTypesLoader,
        ) => {
            return createGraphQLOptions(i18nService, configService, typesLoader, getOptions(configService));
        },
        inject: [ConfigService, I18nService, GraphQLTypesLoader],
        imports: [ConfigModule, I18nModule],
    });
}

function createGraphQLOptions(
    i18nService: I18nService,
    configService: ConfigService,
    typesLoader: GraphQLTypesLoader,
    options: GraphQLApiOptions,
): GqlModuleOptions {
    // Prevent `Type "Node" is missing a "resolveType" resolver.` warnings.
    // See https://github.com/apollographql/apollo-server/issues/1075
    const dummyResolveType = {
        __resolveType() {
            return null;
        },
    };

    return {
        path: '/' + options.apiPath,
        typeDefs: createTypeDefs(),
        include: [options.resolverModule],
        resolvers: {
            JSON: GraphQLJSON,
            DateTime: GraphQLDateTime,
            Node: dummyResolveType,
            PaginatedList: dummyResolveType,
            Upload: GraphQLUpload || dummyResolveType,
        },
        uploads: {
            maxFileSize: configService.assetOptions.uploadMaxFileSize,
        },
        playground: true,
        debug: true,
        context: (req: any) => req,
        extensions: [() => new TranslateErrorExtension(i18nService)],
        // This is handled by the Express cors plugin
        cors: false,
    };

    /**
     * Generates the server's GraphQL schema by combining:
     * 1. the default schema as defined in the source .graphql files specified by `typePaths`
     * 2. any custom fields defined in the config
     * 3. any schema extensions defined by plugins
     */
    function createTypeDefs(): string {
        const customFields = configService.customFields;
        const typeDefs = typesLoader.mergeTypesByPaths(...options.typePaths);
        let schema = generateListOptions(typeDefs);
        schema = addGraphQLCustomFields(schema, customFields);
        const pluginTypes = configService.plugins
            .map(p => (p.defineGraphQlTypes ? p.defineGraphQlTypes() : undefined))
            .filter(notNullOrUndefined);
        for (const types of pluginTypes) {
            schema = extendSchema(schema, types);
        }
        return printSchema(schema);
    }
}
