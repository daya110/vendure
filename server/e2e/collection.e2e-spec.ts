import gql from 'graphql-tag';
import path from 'path';

import {
    CREATE_COLLECTION,
    GET_ASSET_LIST,
    GET_COLLECTION,
    MOVE_COLLECTION,
    UPDATE_COLLECTION,
} from '../../admin-ui/src/app/data/definitions/product-definitions';
import {
    Collection,
    CreateCollection,
    GetAssetList,
    GetCollection,
    LanguageCode,
    MoveCollection,
    SortOrder,
    UpdateCollection,
} from '../../shared/generated-types';
import { ROOT_CATEGORY_NAME } from '../../shared/shared-constants';

import { TEST_SETUP_TIMEOUT_MS } from './config/test-config';
import { TestAdminClient } from './test-client';
import { TestServer } from './test-server';
import { assertThrowsWithMessage } from './test-utils';

describe('Collection resolver', () => {
    const client = new TestAdminClient();
    const server = new TestServer();
    let assets: GetAssetList.Items[];
    let electronicsCategory: Collection.Fragment;
    let laptopsCategory: Collection.Fragment;
    let appleCategory: Collection.Fragment;

    beforeAll(async () => {
        const token = await server.init({
            productsCsvPath: path.join(__dirname, 'fixtures/e2e-products-full.csv'),
            customerCount: 1,
        });
        await client.init();
        const assetsResult = await client.query<GetAssetList.Query, GetAssetList.Variables>(GET_ASSET_LIST, {
            options: {
                sort: {
                    name: SortOrder.ASC,
                },
            },
        });
        assets = assetsResult.assets.items;
    }, TEST_SETUP_TIMEOUT_MS);

    afterAll(async () => {
        await server.destroy();
    });

    describe('createCollection', () => {
        it('creates a root collection', async () => {
            const result = await client.query<CreateCollection.Mutation, CreateCollection.Variables>(
                CREATE_COLLECTION,
                {
                    input: {
                        assetIds: [assets[0].id, assets[1].id],
                        featuredAssetId: assets[1].id,
                        facetValueIds: ['T_1'],
                        translations: [
                            { languageCode: LanguageCode.en, name: 'Electronics', description: '' },
                        ],
                    },
                },
            );

            electronicsCategory = result.createCollection;
            expect(electronicsCategory).toMatchSnapshot();
            expect(electronicsCategory.parent.name).toBe(ROOT_CATEGORY_NAME);
        });

        it('creates a nested category', async () => {
            const result = await client.query<CreateCollection.Mutation, CreateCollection.Variables>(
                CREATE_COLLECTION,
                {
                    input: {
                        parentId: electronicsCategory.id,
                        translations: [{ languageCode: LanguageCode.en, name: 'Laptops', description: '' }],
                        facetValueIds: ['T_2'],
                    },
                },
            );
            laptopsCategory = result.createCollection;
            expect(laptopsCategory.parent.name).toBe(electronicsCategory.name);
        });

        it('creates a 2nd level nested category', async () => {
            const result = await client.query<CreateCollection.Mutation, CreateCollection.Variables>(
                CREATE_COLLECTION,
                {
                    input: {
                        parentId: laptopsCategory.id,
                        translations: [{ languageCode: LanguageCode.en, name: 'Apple', description: '' }],
                        facetValueIds: ['T_3', 'T_4'],
                    },
                },
            );
            appleCategory = result.createCollection;
            expect(appleCategory.parent.name).toBe(laptopsCategory.name);
        });
    });

    describe('collection query', () => {
        it('returns a category', async () => {
            const result = await client.query<GetCollection.Query, GetCollection.Variables>(GET_COLLECTION, {
                id: laptopsCategory.id,
            });
            if (!result.collection) {
                fail(`did not return the category`);
                return;
            }
            expect(result.collection.id).toBe(laptopsCategory.id);
        });

        it('resolves descendantFacetValues 1 level deep', async () => {
            const result = await client.query(GET_DECENDANT_FACET_VALUES, { id: laptopsCategory.id });
            if (!result.collection) {
                fail(`did not return the category`);
                return;
            }
            expect(result.collection.descendantFacetValues.map(v => v.id)).toEqual(['T_3', 'T_4']);
        });

        it('resolves descendantFacetValues 2 levels deep', async () => {
            const result = await client.query(GET_DECENDANT_FACET_VALUES, { id: electronicsCategory.id });
            if (!result.collection) {
                fail(`did not return the category`);
                return;
            }
            expect(result.collection.descendantFacetValues.map(v => v.id)).toEqual(['T_2', 'T_3', 'T_4']);
        });

        it('resolves ancestorFacetValues at root', async () => {
            const result = await client.query(GET_ANCESTOR_FACET_VALUES, { id: electronicsCategory.id });
            if (!result.collection) {
                fail(`did not return the category`);
                return;
            }
            expect(result.collection.ancestorFacetValues.map(v => v.id)).toEqual([]);
        });

        it('resolves ancestorFacetValues 1 level deep', async () => {
            const result = await client.query(GET_ANCESTOR_FACET_VALUES, { id: laptopsCategory.id });
            if (!result.collection) {
                fail(`did not return the category`);
                return;
            }
            expect(result.collection.ancestorFacetValues.map(v => v.id)).toEqual(['T_1']);
        });

        it('resolves ancestorFacetValues 2 levels deep', async () => {
            const result = await client.query(GET_ANCESTOR_FACET_VALUES, { id: appleCategory.id });
            if (!result.collection) {
                fail(`did not return the category`);
                return;
            }
            expect(result.collection.ancestorFacetValues.map(v => v.id)).toEqual(['T_1', 'T_2']);
        });
    });

    describe('updateCollection', () => {
        it('updates the details', async () => {
            const result = await client.query<UpdateCollection.Mutation, UpdateCollection.Variables>(
                UPDATE_COLLECTION,
                {
                    input: {
                        id: appleCategory.id,
                        assetIds: [assets[1].id],
                        featuredAssetId: assets[1].id,
                        facetValueIds: ['T_3'],
                        translations: [{ languageCode: LanguageCode.en, description: 'Apple stuff ' }],
                    },
                },
            );

            expect(result.updateCollection).toMatchSnapshot();
        });
    });

    describe('moveCollection', () => {
        it('moves a category to a new parent', async () => {
            const result = await client.query<MoveCollection.Mutation, MoveCollection.Variables>(
                MOVE_COLLECTION,
                {
                    input: {
                        categoryId: appleCategory.id,
                        parentId: electronicsCategory.id,
                        index: 0,
                    },
                },
            );

            expect(result.moveCollection.parent.id).toBe(electronicsCategory.id);

            const positions = await getChildrenOf(electronicsCategory.id);
            expect(positions.map(i => i.id)).toEqual([appleCategory.id, laptopsCategory.id]);
        });

        it('alters the position in the current parent', async () => {
            await client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                input: {
                    categoryId: appleCategory.id,
                    parentId: electronicsCategory.id,
                    index: 1,
                },
            });

            const afterResult = await getChildrenOf(electronicsCategory.id);
            expect(afterResult.map(i => i.id)).toEqual([laptopsCategory.id, appleCategory.id]);
        });

        it('corrects an out-of-bounds negative index value', async () => {
            await client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                input: {
                    categoryId: appleCategory.id,
                    parentId: electronicsCategory.id,
                    index: -3,
                },
            });

            const afterResult = await getChildrenOf(electronicsCategory.id);
            expect(afterResult.map(i => i.id)).toEqual([appleCategory.id, laptopsCategory.id]);
        });

        it('corrects an out-of-bounds positive index value', async () => {
            await client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                input: {
                    categoryId: appleCategory.id,
                    parentId: electronicsCategory.id,
                    index: 10,
                },
            });

            const afterResult = await getChildrenOf(electronicsCategory.id);
            expect(afterResult.map(i => i.id)).toEqual([laptopsCategory.id, appleCategory.id]);
        });

        it(
            'throws if attempting to move into self',
            assertThrowsWithMessage(
                () =>
                    client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                        input: {
                            categoryId: appleCategory.id,
                            parentId: appleCategory.id,
                            index: 0,
                        },
                    }),
                `Cannot move a Collection into itself`,
            ),
        );

        it(
            'throws if attempting to move into a decendant of self',
            assertThrowsWithMessage(
                () =>
                    client.query<MoveCollection.Mutation, MoveCollection.Variables>(MOVE_COLLECTION, {
                        input: {
                            categoryId: appleCategory.id,
                            parentId: appleCategory.id,
                            index: 0,
                        },
                    }),
                `Cannot move a Collection into itself`,
            ),
        );

        async function getChildrenOf(parentId: string): Promise<Array<{ name: string; id: string }>> {
            const result = await client.query(GET_CATEGORIES);
            return result.collections.items.filter(i => i.parent.id === parentId);
        }
    });
});

const GET_CATEGORIES = gql`
    query GetCategories {
        collections(languageCode: en) {
            items {
                id
                name
                position
                parent {
                    id
                    name
                }
            }
        }
    }
`;

const GET_DECENDANT_FACET_VALUES = gql`
    query GetDescendantFacetValues($id: ID!) {
        collection(id: $id) {
            id
            descendantFacetValues {
                id
                name
            }
        }
    }
`;

const GET_ANCESTOR_FACET_VALUES = gql`
    query GetAncestorFacetValues($id: ID!) {
        collection(id: $id) {
            id
            ancestorFacetValues {
                id
                name
            }
        }
    }
`;
