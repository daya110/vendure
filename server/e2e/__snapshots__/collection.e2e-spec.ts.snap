// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection resolver createCollection creates a root collection 1`] = `
Object {
  "assets": Array [
    Object {
      "fileSize": 4,
      "id": "T_1",
      "mimeType": "image/jpeg",
      "name": "derick-david-409858-unsplash.jpg",
      "preview": "test-url/test-assets/derick-david-409858-unsplash__preview.jpg",
      "source": "test-url/test-assets/derick-david-409858-unsplash.jpg",
      "type": "IMAGE",
    },
    Object {
      "fileSize": 4,
      "id": "T_2",
      "mimeType": "image/jpeg",
      "name": "alexandru-acea-686569-unsplash.jpg",
      "preview": "test-url/test-assets/alexandru-acea-686569-unsplash__preview.jpg",
      "source": "test-url/test-assets/alexandru-acea-686569-unsplash.jpg",
      "type": "IMAGE",
    },
  ],
  "children": null,
  "description": "",
  "featuredAsset": Object {
    "fileSize": 4,
    "id": "T_1",
    "mimeType": "image/jpeg",
    "name": "derick-david-409858-unsplash.jpg",
    "preview": "test-url/test-assets/derick-david-409858-unsplash__preview.jpg",
    "source": "test-url/test-assets/derick-david-409858-unsplash.jpg",
    "type": "IMAGE",
  },
  "filters": Array [
    Object {
      "args": Array [
        Object {
          "name": "facetValueIds",
          "type": "FACET_VALUE_IDS",
          "value": "[\\"T_1\\"]",
        },
      ],
      "code": "facet-value-filter",
      "description": "Filter by FacetValues",
    },
  ],
  "id": "T_3",
  "languageCode": "en",
  "name": "Electronics",
  "parent": Object {
    "id": "T_1",
    "name": "__root_collection__",
  },
  "translations": Array [
    Object {
      "description": "",
      "id": "T_3",
      "languageCode": "en",
      "name": "Electronics",
    },
  ],
}
`;

exports[`Collection resolver updateCollection 1`] = `
Object {
  "assets": Array [
    Object {
      "fileSize": 4,
      "id": "T_1",
      "mimeType": "image/jpeg",
      "name": "derick-david-409858-unsplash.jpg",
      "preview": "test-url/test-assets/derick-david-409858-unsplash__preview.jpg",
      "source": "test-url/test-assets/derick-david-409858-unsplash.jpg",
      "type": "IMAGE",
    },
  ],
  "children": null,
  "description": "Apple stuff ",
  "featuredAsset": Object {
    "fileSize": 4,
    "id": "T_1",
    "mimeType": "image/jpeg",
    "name": "derick-david-409858-unsplash.jpg",
    "preview": "test-url/test-assets/derick-david-409858-unsplash__preview.jpg",
    "source": "test-url/test-assets/derick-david-409858-unsplash.jpg",
    "type": "IMAGE",
  },
  "filters": Array [
    Object {
      "args": Array [
        Object {
          "name": "facetValueIds",
          "type": "FACET_VALUE_IDS",
          "value": "[\\"T_3\\"]",
        },
      ],
      "code": "facet-value-filter",
      "description": "Filter by FacetValues",
    },
  ],
  "id": "T_5",
  "languageCode": "en",
  "name": "Pear",
  "parent": Object {
    "id": "T_4",
    "name": "Computers",
  },
  "translations": Array [
    Object {
      "description": "Apple stuff ",
      "id": "T_5",
      "languageCode": "en",
      "name": "Pear",
    },
  ],
}
`;
